webrtc/README.md

```diff
@@ -0,0 +1,130 @@
+# WebRTC
+
+| Lifecycle Stage | Maturity      | Status | Latest Revision |
+|-----------------|---------------|--------|-----------------|
+| 1A              | Working Draft | Active |                 |
+
+Authors: [@mxinden]
+
+Interest Group: [@marten-seemann]
+
+## Motivation
+
+1. **No need for valid TLS certificates.** Enable browsers to connect to public
+   server nodes without those server nodes providing a TLS certificate within
+   the browsers trustchain. Note that we can not do this today with our
+   Websocket transport.
+
+2. **Hole punching in the browser**: Enable two browsers or a browser and a
+   non-public server node to connect.
+
+## Requirements
+
+- Loading a remote nodes certificate into ones browser trust-store is not an
+  option, i.e. doesn't scale.
+
+- No dependency on central STUN and/or TURN servers.
+
+## Protocol
+
+Scenario: Browser _A_ wants to connect to server node _B_.
+
+### Setup - Both _A_ and _B_
+
+1. [Generate a
+   certificate](https://www.w3.org/TR/webrtc/#dom-rtcpeerconnection-generatecertificate).
+
+2. [Get the certificate's
+   fingerprint](https://www.w3.org/TR/webrtc/#dom-rtccertificate-getfingerprints).
+
+3. Sign this fingerprint using the libp2p peer key to generate a "peer cert"
+
+### Connection Establishment
+
+1. Browser _A_ discovers server node _B_'s multiaddr, containing _B_'s IP, UDP
+  port, TLS certificate fingerprint and libp2p peer ID (e.g.
+  `/ip6/2001:db8::/udp/1234/webrtc/<tls-certificate-fingerprint/p2p/<peer-id>`),
```

Comment:
Interesting point here @tomaka. To better understand your comment I will attempt to summarize the options we have so far:
1) Unique persistent certificates that are stored similarly to PeerId
2) Same persistent certificate that is hard coded into Libp2p
Is that correct? 

IMO, I think that 1) makes the most sense taking browsers into consideration. Browsers uniquely generate new certificates in every new session, so consistency is nice.
However 2) is quite interesting too. If we have 2) instead of `webrtc/<tls-cert>`, we could distinguish between browser like nodes and server like nodes (although I'm not sure if that is appropriate to distinguish node types in MultiAddr)/

Reply:
These are indeed the two solutions. The drawback of solution 2 is that the DTLS session basically doesn't protect you against anything, so you need to open a single WebRTC data stream and run an extra encryption and multiplexing layer within that data stream, which adds an overhead.

> If we have 2) instead of webrtc/<tls-cert>, we could distinguish between browser like nodes and server like nodes (although I'm not sure if that is appropriate to distinguish node types in MultiAddr)/

In my opinion, we should have two different multiaddress protocols: one that looks like `/webrtc/<tls-cert>` and one that looks like `/webrtc-hardcoded`.
In order to connect to a `/webrtc/<tls-cert>` address, you do as this RFC presently described, and in order to connect to a `/webrtc-hardcoded` address, you use the hardcoded certificate and open a data stream and negotiate noise and yamux.

Reply:
> This is already more or less true for the peer id, which is not possible to determine manually.

why? if the public key is stored somewhere like `{blockchain_node}/.peer_id/public_key.pub`, then you can write a script (or even have a subcommand like `{your framework} peer_id`, which reads the public key and calculates peer_id, no?

Same with the certificate. If the location is commonly agreed upon, then you can calculate sha-256 of the certificate.

It does put additional burden on devops / framework developers though.

Reply:
I believe the complexity (the need to persist both libp2p key pair and TLS certificate material) is worth not having to double-encrypt (DTLS + Noise) and double multiplex (SCTP + Yamux).

That said, I agree that multiple WebRTC styles can co-exist. Though I would like to push for the one proposed in this specification first.

> However the peer id is not strictly mandatory (as opposed to the TLS certificate fingerprint)

For the sake of completeness, while not mandatory (in some implementations) to establish a connection, it is mandatory to establish a secured connection.

Reply:
> I believe the complexity (the need to persist both libp2p key pair and TLS certificate material) is worth not having to double-encrypt (DTLS + Noise) and double multiplex (SCTP + Yamux).

I believe the opposite. Users of libp2p typically have a problem of complexity more than they have a problem of CPU usage.

> while not mandatory (in some implementations) to establish a connection, it is mandatory to establish a secured connection.

I wouldn't say that. Even if you don't know the PeerId in advance, the connection you establish is still encrypted.
It is true that if you connect to an IP address without a PeerId you might not actually connect to said IP address but to a different one, but I have trouble seeing a situation where this is problematic. IP addresses shouldn't be used as identity anyway.

Reply:
This is perhaps a niche use-case, but for a project I'm working on, I would like to be able to re-use the libp2p webrtc peer connection to add additional unreliable data channels and media tracks.

If the DTLS connection is not trusted to be secure, then every data channel would need to be Noise encrypted, and any media tracks would insecure.

It's also worth keeping in mind that any additional in-browser encryption/multiplexing is going to be happening in single-threaded wasm. I think it'd be worth profiling to see what kind of overhead that adds.

Reply:
I just want to repeat my point:
I'm advocating for giving both options, having a multiaddr protocol where the certificate fingerprint in the multiaddr and using DTLS/SCTP for the encryption & multiplexing (as currently explained in this spec), but also having a separate multiaddr protocol that uses a hardcoded certificate, opens a single data stream, and treats it as an unsecure channel.

I also want to point out that it would in no way be mandatory for an implementation of libp2p to support both multiaddr protocols.

Reply:
"Hardcoded certificate" sounds like a euphemism, what we're actually talking about is publishing a private key.
This makes me feel uncomfortable. Of course, running all communication via a single stream (with a TCP stream muxer applied) would secure that one stream, but this comes with a multitude of problems:
1. We need two completely different WebRTC session implementations, as `webrtc` will use WebRTC streams.
2. It's not clear what happens when an attacker opens another stream. At the minimum, it can mess with flow control.
3. It precludes the use case that @expenses mentioned of using the audio / video / unreliable data channels that the bare WebRTC connection gives you.

I'm not convinced that this statement is true to begin with:
> which is that from a practical point of view you can't manually figure out the multiaddress of a node that you set up

This is only true if you generate a random certificate. But that's not your only option. You can (HKDF-) extract predictable, cryptographically secure pseudo-randomness from your node's private key, thereby giving you a deterministic, unique certificate.

Reply:
> We need two completely different WebRTC session implementations, as webrtc will use WebRTC streams.

I don't think that's a problem? Most of the code would be shared between the two WebRTC protocols, and the difference between the two are relatively minor.

> It precludes the use case that @expenses mentioned of using the audio / video / unreliable data channels that the bare WebRTC connection gives you.

No it does not. @expenses can use the protocol with multiple data streams.
Again, I'm asking for having both options, and not replacing the current one with hardcoded certificate.

> It's not clear what happens when an attacker opens another stream. At the minimum, it can mess with flow control.

That's a valid argument. I don't know enough about how DTLS works, but considering that an attacker knows the cipher state they can send a packet to the browser that the browser thinks legitimately comes from the server. While this invalid packet will not be able to pass through our Noise layer, it will corrupt the cipher in the browser and lead to legitimate packets being considered as invalid.

Reply:
However, considering that the client would still use a properly generated certificate, tempering with a connection would require eavesdropping the initial DTLS handshake, which means being on the path of the packets. Routers on the path of the packets can already if I'm not mistaken kill TCP connections if they wish so by sending fake packets. In other words, I don't think this is more a problem than with a regular TCP/IP+TLS connection.

Reply:
Again, I don't think the premise that motivates this hack is correct. In fact, you trivially can, as I've described in as I've described in https://github.com/libp2p/specs/pull/412#discussion_r882768879, know the certificate hash in advance, if you derive the DTLS private key from your host key.

> I don't think that's a problem? Most of the code would be shared between the two WebRTC protocols, and the difference between the two are relatively minor.

There's a cost (maintenance, understandability, security) to complexity, both in specification and implementation.

> No it does not. @expenses can use the protocol with multiple data streams.

Things get more complicated with unreliable data channels. You'd have to run a DTLS handshake on _every_ channel then, adding computational and latency overhead. As unreliable channels are often used for "fire and forget" semantics, this would be prohibitive.
I'm not familiar with how WebRTC audio and video works, but I highly doubt that browsers would you give an option to double-encrypt the data.

Reply:
> In fact, you trivially can, as I've described in as I've described in https://github.com/libp2p/specs/pull/412#discussion_r882768879, know the certificate hash in advance, if you derive the DTLS private key from your host key.

This is of course also a solution and was actually [my first approach](https://github.com/tomaka/substrate-lite-1/blob/be4d077596bb95f3ff67e8724363c96f47d02271/bin/wasm-node/javascript/src/webrtc.ts#L85-L90).
I've discarded this solution because I don't think it's that trivial and that the hardcoded certificate seems easier.

Certificates are big data structures, and we would have to precisely specify the order of the fields and all the small details, a bit like we're doing in [the encoding of a PeerId](https://github.com/libp2p/specs/blob/master/peer-ids/peer-ids.md#keys) but more complicated. You also have to implement this certificate generation in JavaScript for the browser.

> Things get more complicated with unreliable data channels. You'd have to run a DTLS handshake on every channel then, adding computational and latency overhead. As unreliable channels are often used for "fire and forget" semantics, this would be prohibitive.
I'm not familiar with how WebRTC audio and video works, but I highly doubt that browsers would you give an option to double-encrypt the data.

Sorry but I've repeated this multiple times now. **I'm proposing that we give both options: a multiaddr protocol for the hardcoded certificate, and a multiaddr protocol for the "regular" WebRTC protocol.** @expenses's use case would use exclusively the latter. **@expenses's use case is completely irrelevant to the discussion in this comment.**

Reply:
Considering that everyone seems to ignore my concern, I will just shut up.
For years now I've been spending a considerable amount of time answering questions from node operators who are confused about peer ids and networking protocol, and I am very confident in the fact that the necessity to provide a certificate hash is going to be a shit show in terms of complexity for end users (which to me matters more than extra complexity in the specs), but I am unwilling to spend more energy on this.

Reply:
> Sorry but I've repeated this multiple times now. **I'm proposing that we give both options: a multiaddr protocol for the hardcoded certificate, and a multiaddr protocol for the "regular" WebRTC protocol.**

While the hardcoded certificate protocol would be much easier to implement, I'm sure that having two seperate incompatible protocols for WebRTC would confused a lot of people who aren't in the know as well. There are no good options here and it might take some experimenting to find out which one is least bad.

Reply:
> Considering only the browser to server case, would it be possible to establish an insecure websocket connection and then negotiate NOISE for SDP and candidate exchange?

This is problematic for multiple reasons:
1. Browsers won’t allow insecure WebSocket connections on websites loaded via HTTPS.
2. Even if they did, once we have a WebSocket connection, we can just upgrade that one (run the libp2p handshake and apply a stream muxer), so there’s no real need for WebRTC.

Reply:
Got it, thanks! I would assume we would want to use websocket and WebRTC connections for different purposes though.

Reply:
Yes @ckousik. Maybe https://github.com/libp2p/specs/pull/422 helps paint the greater picture?

Reply:
I am resolving here. This specification will not go the route of hard-coded certificates. That is not to say that another specification can not go down that route.